### ISSUE TITLE
Hardcoded dependency to SonataMediaBundle in Model/Category

### ISSUE THREAD

**Issue by soerenbernstein** [sonata-project/SonataClassificationBundle#269]
#### Sonata packages
sonata-project/admin-bundle              3.11.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.3.0  Symfony SonataBlockBundle
sonata-project/cache                     1.0.7  Cache library
sonata-project/classification-bundle     3.2.0  Symfony SonataClassificationBundle
sonata-project/core-bundle               3.2.0  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.2    Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.1.3  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.1.10 Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.7.0  Lightweight Exporter library
sonata-project/google-authenticator      1.0.2  Library to integrate Google Authenticator into a PHP project
sonata-project/intl-bundle               2.2.4  Symfony SonataIntlBundle
sonata-project/user-bundle               3.2.0  Symfony SonataUserBundle

#### Symfony packages
symfony/assetic-bundle             v2.8.1  Integrates Assetic into Symfony2
symfony/framework-standard-edition v2.7.23 The "Symfony Standard Edition" distribution
symfony/monolog-bundle             v3.0.3  Symfony MonologBundle
symfony/polyfill-apcu              v1.3.0  Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-mbstring          v1.3.0  Symfony polyfill for the Mbstring extension
symfony/polyfill-php54             v1.3.0  Symfony polyfill backporting some PHP 5.4+ features to lower PHP versions
symfony/polyfill-php55             v1.3.0  Symfony polyfill backporting some PHP 5.5+ features to lower PHP versions
symfony/polyfill-php56             v1.3.0  Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70             v1.3.0  Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util              v1.3.0  Symfony utilities for portability of PHP codes
symfony/swiftmailer-bundle         v2.4.2  Symfony SwiftmailerBundle
symfony/symfony                    v2.7.23 The Symfony PHP framework

#### PHP version
PHP 7.0.10 (cli) (built: Aug 31 2016 16:01:36) ( NTS )

## Subject
The type hint of parameter "media" in method "setMedia" in class "Sonata\ClassificationBundle\Entity\Category" is invalid.

\Sonata\ClassificationBundle\Model\Category uses an Interface from MediaBundle, which is an optional dependency and not installed in my case.

**Comment by netandreus** [comment]
Have same issue on sonata-project/classification-bundle 3.2.1 and 3.2.0

**Comment by greg0ire** [comment]
Can you provide a stack trace? I wonder when the error occurs.

**Comment by soerenbernstein** [comment]
No, `setMedia` is not called, as far as I am aware. But the method signature contains a type hint to `MediaInterface,` which is not part of the classification bundle, but the media bundle. The media bundle is not installed, so the Interface does not exists, so php throws an error.

This is also true for \Sonata\ClassificationBundle\Model\Collection.

**Comment by ZRazor** [comment]
I created a test project, and everything is ok
<img width="692" alt="2017-03-01 18 42 06" src="https://cloud.githubusercontent.com/assets/1560126/23452171/0b34ab56-feaf-11e6-8d0c-f2f2dca9270e.png">

Composer:
```
        "symfony/symfony": "2.7.23",
        "doctrine/orm": "^2.4.8",
        "doctrine/doctrine-bundle": "~1.4",
        "symfony/assetic-bundle": "~2.3",
        "symfony/swiftmailer-bundle": "~2.3,>=2.3.10",
        "symfony/monolog-bundle": "^3.0.2",
        "sensio/distribution-bundle": "~4.0",
        "sensio/framework-extra-bundle": "^3.0.2",
        "incenteev/composer-parameter-handler": "~2.0",
        "sonata-project/admin-bundle": "3.11.0",
        "sonata-project/block-bundle": "3.3.0",
        "sonata-project/cache": "1.0.7",
        "sonata-project/classification-bundle": "3.2.0",
        "sonata-project/datagrid-bundle": "2.2",
        "sonata-project/doctrine-extensions": "1.0.2",
        "sonata-project/doctrine-orm-admin-bundle": "3.1.3",
        "sonata-project/easy-extends-bundle": "2.1.10",
        "sonata-project/exporter": "1.7.0",
        "sonata-project/intl-bundle": "2.2.4"
```
PHP: `PHP 7.0.15 (cli) (built: Feb 23 2017 16:49:41) ( NTS )`

@soerenbernstein, can you provide more info?

**Comment by ZRazor** [comment]
Suddenly got this error
```
The type hint of parameter "media" in method "setMedia" in class "Sonata\ClassificationBundle\Model\Category" is invalid.
```

Full stack trace:
```
Doctrine\Common\Proxy\Exception\UnexpectedValueException: The type hint of parameter "media" in method "setMedia" in class "Sonata\ClassificationBundle\Model\Category" is invalid.
    at n/a
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/Exception/UnexpectedValueException.php line 57

    at Doctrine\Common\Proxy\Exception\UnexpectedValueException::invalidParameterTypeHint('Sonata\ClassificationBundle\Model\Category', 'setMedia', 'media')
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/ProxyGenerator.php line 1071

    at Doctrine\Common\Proxy\ProxyGenerator->formatType(object(ReflectionType), object(ReflectionMethod), object(ReflectionParameter))
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/ProxyGenerator.php line 947

    at Doctrine\Common\Proxy\ProxyGenerator->getParameterType(object(ClassMetadata), object(ReflectionMethod), object(ReflectionParameter))
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/ProxyGenerator.php line 908

    at Doctrine\Common\Proxy\ProxyGenerator->buildParametersString(object(ClassMetadata), object(ReflectionMethod), array(object(ReflectionParameter)))
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/ProxyGenerator.php line 780

    at Doctrine\Common\Proxy\ProxyGenerator->generateMethods(object(ClassMetadata))
        in  line 

    at call_user_func(array(object(ProxyGenerator), 'generatemethods'), object(ClassMetadata))
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/ProxyGenerator.php line 276

    at Doctrine\Common\Proxy\ProxyGenerator->generateProxyClass(object(ClassMetadata), '/Users/anton/sites/test_bug/app/cache/dev/doctrine/orm/Proxies/__CG__ApplicationSonataClassificationBundleEntityCategory.php')
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php line 220

    at Doctrine\Common\Proxy\AbstractProxyFactory->getProxyDefinition('Application\Sonata\ClassificationBundle\Entity\Category')
        in /Users/anton/sites/test_bug/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php line 119

    at Doctrine\Common\Proxy\AbstractProxyFactory->getProxy('Application\Sonata\ClassificationBundle\Entity\Category', array('id' => '101'))
        in /Users/anton/sites/test_bug/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php line 2707

    at Doctrine\ORM\UnitOfWork->createEntity('Application\Sonata\ClassificationBundle\Entity\Category', array('name' => 'inner_test', 'enabled' => true, 'slug' => 'inner-test', 'description' => null, 'position' => '0', 'createdAt' => object(DateTime), 'updatedAt' => object(DateTime), 'id' => '3', 'parent_id' => '101', 'context' => 'default'), array('deferEagerLoad' => true, 'fetchAlias' => 'c', 'fetchMode' => array('Application\Sonata\ClassificationBundle\Entity\Category' => array('parent' => '2'))))
        in /Users/anton/sites/test_bug/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php line 273

    at Doctrine\ORM\Internal\Hydration\ObjectHydrator->getEntity(array('name' => 'inner_test', 'enabled' => true, 'slug' => 'inner-test', 'description' => null, 'position' => '0', 'createdAt' => object(DateTime), 'updatedAt' => object(DateTime), 'id' => '3', 'parent_id' => '101', 'context' => 'default'), 'c')
        in /Users/anton/sites/test_bug/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php line 494

    at Doctrine\ORM\Internal\Hydration\ObjectHydrator->hydrateRowData(array('name_0' => 'inner_test', 'enabled_1' => true, 'slug_2' => 'inner-test', 'description_3' => null, 'position_4' => '0', 'created_at_5' => '2017-03-02 14:29:44', 'updated_at_6' => '2017-03-02 14:29:44', 'id_7' => '3', 'parent_id_8' => '101', 'context_9' => 'default'), array())
        in /Users/anton/sites/test_bug/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php line 165

    at Doctrine\ORM\Internal\Hydration\ObjectHydrator->hydrateAllData()
        in /Users/anton/sites/test_bug/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php line 147

    at Doctrine\ORM\Internal\Hydration\AbstractHydrator->hydrateAll(object(PDOStatement), object(ResultSetMapping), array())
        in /Users/anton/sites/test_bug/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php line 978

    at Doctrine\ORM\AbstractQuery->executeIgnoreQueryCache(null, null)
        in /Users/anton/sites/test_bug/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php line 924

    at Doctrine\ORM\AbstractQuery->execute()
        in /Users/anton/sites/test_bug/vendor/sonata-project/classification-bundle/Entity/CategoryManager.php line 197

    at Sonata\ClassificationBundle\Entity\CategoryManager->loadCategories(object(Context))
        in /Users/anton/sites/test_bug/vendor/sonata-project/classification-bundle/Entity/CategoryManager.php line 108

    at Sonata\ClassificationBundle\Entity\CategoryManager->getRootCategory(object(Context))
        in /Users/anton/sites/test_bug/vendor/sonata-project/classification-bundle/Form/Type/CategorySelectorType.php line 80

    at Sonata\ClassificationBundle\Form\Type\CategorySelectorType->getChoices(object(OptionsResolver))
        in /Users/anton/sites/test_bug/vendor/sonata-project/classification-bundle/Form/Type/CategorySelectorType.php line 61

    at Sonata\ClassificationBundle\Form\Type\CategorySelectorType->Sonata\ClassificationBundle\Form\Type\{closure}(object(OptionsResolver), null)
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/OptionsResolver/OptionsResolver.php line 859

    at Symfony\Component\OptionsResolver\OptionsResolver->offsetGet('choice_list')
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/OptionsResolver/OptionsResolver.php line 791

    at Symfony\Component\OptionsResolver\OptionsResolver->resolve(array('sonata_field_description' => object(FieldDescription), 'category' => object(Category), 'model_manager' => object(ModelManager), 'class' => 'Application\Sonata\ClassificationBundle\Entity\Category', 'required' => true, 'context' => object(Context), 'label_render' => false, 'label' => 'form.label_parent'))
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/Form/ResolvedFormType.php line 156

    at Symfony\Component\Form\ResolvedFormType->createBuilder(object(FormFactory), 'parent', array('sonata_field_description' => object(FieldDescription), 'category' => object(Category), 'model_manager' => object(ModelManager), 'class' => 'Application\Sonata\ClassificationBundle\Entity\Category', 'required' => true, 'context' => object(Context), 'label_render' => false, 'label' => 'form.label_parent'))
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/Form/Extension/DataCollector/Proxy/ResolvedTypeDataCollectorProxy.php line 89

    at Symfony\Component\Form\Extension\DataCollector\Proxy\ResolvedTypeDataCollectorProxy->createBuilder(object(FormFactory), 'parent', array('sonata_field_description' => object(FieldDescription), 'category' => object(Category), 'model_manager' => object(ModelManager), 'class' => 'Application\Sonata\ClassificationBundle\Entity\Category', 'required' => true, 'context' => object(Context), 'label_render' => false, 'label' => 'form.label_parent'))
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/Form/FormFactory.php line 114

    at Symfony\Component\Form\FormFactory->createNamedBuilder('parent', object(ResolvedTypeDataCollectorProxy), null, array('sonata_field_description' => object(FieldDescription), 'category' => object(Category), 'model_manager' => object(ModelManager), 'class' => 'Application\Sonata\ClassificationBundle\Entity\Category', 'required' => true, 'context' => object(Context), 'label_render' => false, 'label' => 'form.label_parent'))
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php line 106

    at Symfony\Component\Form\FormBuilder->create('parent', 'sonata_category_selector', array('sonata_field_description' => object(FieldDescription), 'category' => object(Category), 'model_manager' => object(ModelManager), 'class' => 'Application\Sonata\ClassificationBundle\Entity\Category', 'required' => true, 'context' => object(Context), 'label_render' => false, 'label' => 'form.label_parent'))
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php line 269

    at Symfony\Component\Form\FormBuilder->resolveChildren()
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php line 215

    at Symfony\Component\Form\FormBuilder->getForm()
        in /Users/anton/sites/test_bug/vendor/sonata-project/admin-bundle/Admin/AbstractAdmin.php line 3198

    at Sonata\AdminBundle\Admin\AbstractAdmin->buildForm()
        in /Users/anton/sites/test_bug/vendor/sonata-project/admin-bundle/Admin/AbstractAdmin.php line 1324

    at Sonata\AdminBundle\Admin\AbstractAdmin->getForm()
        in /Users/anton/sites/test_bug/vendor/sonata-project/admin-bundle/Controller/CRUDController.php line 496

    at Sonata\AdminBundle\Controller\CRUDController->createAction()
        in  line 

    at call_user_func_array(array(object(CategoryAdminController), 'createAction'), array())
        in /Users/anton/sites/test_bug/app/bootstrap.php.cache line 3238

    at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), '1')
        in /Users/anton/sites/test_bug/app/bootstrap.php.cache line 3197

    at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), '1', true)
        in /Users/anton/sites/test_bug/app/bootstrap.php.cache line 3351

    at Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel->handle(object(Request), '1', true)
        in /Users/anton/sites/test_bug/app/bootstrap.php.cache line 2536

    at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
        in /Users/anton/sites/test_bug/web/app_dev.php line 29

    at require('/Users/anton/sites/test_bug/web/app_dev.php')
        in /Users/anton/sites/test_bug/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Resources/config/router_dev.php line 40
```

**Comment by soerenbernstein** [comment]
There you go, that's the problem. :-)

**Comment by micayael** [comment]
Is there a fix for this?

**Comment by lracicot** [comment]
@OskarStark Would it be possible to confirm the issue? Should we work on a PR to improve the doc on this regards or to remove the dependency?

**Comment by soerenbernstein** [comment]
Well, you should remove the dependency.

**Comment by tonyellow** [comment]
I also have this problem. Im using classfication bundle but not media bundle and it is throwing errors because of this dependancy. Is there a nice way to fix this?

**Comment by core23** [comment]
Fixed with #196
